### ISSUE TITLE
Attribute order is not defined

### ISSUE THREAD

**Issue by oznetmaster** [nunit/nunit#2326]
I seem to be constantly bringing this up.

The order of attributes is undefined on all platforms.  GetCustomAttribute(s) can return the attributes in any order that it wants.

Tests that assume a specific order for the attributes will fail on some or all platforms.

One example is the PR for #2079 which add TestMethodTests, many of which assume a specific order for the TestCase attributes.  These tests, for instance will fail on the CF platform.

**Comment by jnm2** [comment]
We should definitely fix that across our tests. Are you up for that?

**Comment by jnm2** [comment]
We should add this to our internal NUnit analyzer bucket list since it's constantly hard to notice.

**Comment by oznetmaster** [comment]
I am leaving on an 8 week vacation where I will have little or no internet access for at least half of it.

I will see where things stand when I return.

**Comment by jnm2** [comment]
All right. Until then, I'll put it up for grabs. What we're looking for is any test where we build tests and access the child `.Tests` collection in an order-dependent fashion, or where we run tests and access the results in an order-dependent fashion.

What we should do instead is assume the tests are in random order and compare accordingly.
I could imagine for example [these lines](https://github.com/nunit/nunit/pull/2096/files#diff-1acc1606be8d10a0a6658d83cb4d5964R69):

```c#
Assert.That(test.Tests[0].Arguments, Is.EqualTo(new object[] { 42, "abc" }));
Assert.That(test.Tests[1].Arguments, Is.EqualTo(new object[] { 24, "cba" }));
```

Being written similar to:
``c#
Assert.That(from child in test.Tests select child.Arguments, Is.EquivalentTo(new[]
{
    new object[] { 42, "abc" },
    new object[] { 24, "cba" }
}));
```

@oznetmaster Please correct me if I'm missing anything.

**Comment by CharliePoole** [comment]
It's not quite that simple. You can assume test order when it's not dependent on attribute order. For example, if TestCaseSource returns 10 cases in a given order, that order won't change. The extent to which order is and isn't guaranteed is well known and even documented. See https://github.com/nunit/docs/wiki/TestCaseSource-Attribute

**Comment by CharliePoole** [comment]
Andi course it's perfectly OK to look at Tests[0] when there is only one test. Let's fix any identifiable problems but not start a big search for more. I'm fairly sure this is not widespread and is likely to be a recent phenomenon.

**Comment by jnm2** [comment]
Also worth pointing out that the order of test methods and the order of test classes is also not defined.

**Comment by CharliePoole** [comment]
@jnm2 Right, but relevant?

**Comment by jnm2** [comment]
As relevant as the test case order problem, no?

**Comment by CharliePoole** [comment]
Question was whether it's relevant __to__ the testcase order problem.

However, let's not split hairs. I just think you may be sending some new contributor on a wild goose chase, looking for a problem we have not confirmed to be present beyond a few cases that @oznetmaster identified. We might better spend time fixing each case in a PR as it comes up.

OTOH, if we really want to spend time looking for these hypothetical problems, it will take someone with a pretty good understanding of how NUnit works.

If we do actually have hidden bugs waiting to hit us, we don't know where they are by definition, so let's swat them when they appear.

**Comment by rprouse** [comment]
Nobody has picked this up and nobody else has asked for this so closing. This might also be fixed by #2863?
